### DISCUSSION TITLE
Microbunching gain and intrabeam scattering calculations

### DISCUSSION THREAD

**Discussion by abryxkc** [ocelot-collab/ocelot#253] [Ideas]
Hi all,

Would it be feasible to include a PhysicsProcess to calculate the microbunching gain step-by-step along the lattice? 
I think that much of the scaffolding is there, as the longitudinal space-charge and coherent synchrotron radiation impedances have been written already. 

I have some scripts which can take the output of an ELEGANT simulation and calculate the gain along the lattice which could be useful in this instance. 

An additional feature that could be of interest (to me, at least) would be the inclusion of intrabeam scattering effects. This is arguably more tricky, as there are assumptions about the beam that may not necessarily hold. It could be interesting nevertheless. 

Let me know if you are interested in working on this. As you guys are more familiar with how the code works in general, your help would be appreciated!

Thanks,
Alex

**Comment by sergey-tomin** [comment]
Hi Alex,

The topic is interesting indeed. If you have already started something or are interested in developing MBI or/and IBS and implement it in Ocelot, you are very welcome. I would be glad to help you. I began working on this but have not had time to make significant progress. How are you calculating MBI gain? A few years ago, I implemented it based on the work of Saldin, Yurkov, and Schneidmiller, but I did not include it in Ocelot. Is your implementation different?
I have a tutorial which explain how to make your own PhysicsProc https://github.com/ocelot-collab/ocelot/blob/master/demos/ipython_tutorials/8_laser_heater.ipynb 

Cheers,
Sergey.

**Comment by abryxkc** [comment]
Great! Thank you for your interest.

My method is based on one from [this paper](https://journals.aps.org/prab/abstract/10.1103/PhysRevAccelBeams.23.124401). 
And others by the same author(s).

Their model has lots of "bells and whistles", including IBS effects, but it is pretty similar to that of Saldin et al, and Huang&Kim for CSR.

So I do tracking in ELEGANT, then:
1. Load in the beam and lattice parameters from the simulation -- energy spread, gamma, beam sizes, R_xx transport functions, etc...
2. Calculate the initial bunching factor based on initial current and energy spread.
3. At each step in the lattice, compute the kernel functions due to LSC/CSR impedance using the beam properties. Do this for a range of initial modulation wavelengths (say between 1 um and 50 um). The simplest model just has the kernels for the collective effects, and you can also include IBS.
4. Add this to the previous value of the bunching factor. 
5. Do this along the entire line (loop back to step 3).
6. The final MBI gain is then bf / b0. The bunching in the energy plane is also calculated. 

Since with a PhysProc you can do these calculations step-by-step, and all the information is already contained in the ParticleArray and the MagneticLattice, this could be done alongside other calculations during the tracking. 
One possible concern is that it could slow things down, since you need to calculate the bunching factor at each step for a range of modulation wavelengths. 

Thanks for sharing the notebook, and your interest. I will read up, and let's discuss some more :)
I tried to attach my scripts in a zip file if you would like to take a look. Not sure if it worked. 
The scripts aren't particularly "tidy", and there may be some bugs (particularly in the IBS kernels, but you can exclude these from the calculation). Usual caveats apply :smile: 
[mbi_calc.tar.gz](https://github.com/ocelot-collab/ocelot/files/15415111/mbi_calc.tar.gz)

Thanks,
Alex

**Comment by abryxkc** [comment]
I should also add another point: the usual way of doing these calculations assumes a uniform beam distribution, which may not be the most correct way.
For example, if you have a beam with a curved longitudinal phase space, you need to be careful about how you define the energy spread. Similarly for other properties that may vary along the beam.

In fact this is one reason why I thought OCELOT would be a good tool for doing this, as you can calculate the slice properties along the beam (again slowing down the calculation but hopefully producing a more accurate result). Hopefully this makes sense.
